src/Voron/Impl/FreeSpace/StreamBitArray.cs

```diff
@@ -110,6 +114,180 @@ public bool Get(int index)
             return (_inner[index >> 5] & (1 << (index & 31))) != 0;
         }
 
+        public int? GetContinuousRangeStart(int num)
+        {
+            switch (num)
+            {
+                case 1:
+                    // finding a single set bit
+                    for (var i = 0; i < _inner.Length; i++)
+                    {
+                        switch (_inner[i])
+                        {
+                            case 0:
+                                continue;
+                            case -1:
+                                return i * BitsInItem;
+                            default:
+                                return i * BitsInItem + BitOperations.TrailingZeroCount(_inner[i]);
+                        }
+                    }
+
+                    return null;
+
+                case < BitsInItem:
+                    // finding sequences up to 32 bits
+
+                    for (var i = 0; i < _inner.Length; i++)
+                    {
+                        int current = _inner[i];
+                        if (current == 0)
+                            continue;
+
+                        if (current == -1)
+                            return i * BitsInItem;
+
+                        int firstSetBitPos = BitOperations.TrailingZeroCount((uint)current);
+
+                        int mask = (1 << num) - 1;
```

Comment:
Can probably use

```suggestion
                        if (firstSetBitPos <= BitsInItem - num)
                        {
                            int mask = ( -1 >>> (BitsInItem - num) ) << firstSetBitPos ;
                            if ( (current & mask) != 0) 
                               return  return i * BitsInItem + bitPos;
                         }
```

Reply:
Note that this doesn't handle the case of finding a partial match (wanting three pages when you have `0b0111011`).
Need to loop to the _next_ match here.

You code doesn't handle it either.

Reply:
Changed